pkg/reconciler/source/crd/crd.go

```diff
@@ -0,0 +1,308 @@
+	crdLister apiextensionsv1beta1.CustomResourceDefinitionLister
+	// apiExtensionsClientSet used to update finalizers on CRDs.
+	apiExtensionsClientSet clientset.Interface
+
+	ogctx context.Context
+	ogcmw configmap.Watcher
+
+	// controllers keeps a map for GVR to dynamically created controllers.
+	controllers map[schema.GroupVersionResource]runningController
+
+	// Synchronization primitives
+	lock     sync.RWMutex
+	onlyOnce sync.Once
+}
+
+// Check that our Reconciler implements controller.Reconciler
+var _ controller.Reconciler = (*Reconciler)(nil)
+
+func (r *Reconciler) Reconcile(ctx context.Context, key string) error {
+
+	// Create controllers map only once.
+	r.onlyOnce.Do(func() {
+		r.controllers = make(map[schema.GroupVersionResource]runningController)
+	})
+
+	// Convert the namespace/name string into a distinct namespace and name.
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		logging.FromContext(ctx).Error("invalid resource key")
+		return nil
+	}
+
+	// Get the CRD resource with this name.
+	original, err := r.crdLister.Get(name)
+	if apierrs.IsNotFound(err) {
+		// The resource may no longer exist, in which case we stop processing.
+		logging.FromContext(ctx).Error("CRD key in work queue no longer exists")
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	// Don't modify the informers copy.
+	crd := original.DeepCopy()
+
+	var reconcileErr error
+	if crd.GetDeletionTimestamp().IsZero() {
+		// add the finalizer
+		if crd, err = r.addFinalizer(ctx, crd); err != nil {
+			logging.FromContext(ctx).Warn("Failed to add finalizers", zap.Error(err))
+		}
+
+		// Reconcile this copy of the resource.
+		reconcileErr = r.reconcile(ctx, crd)
+	} else {
+		// If this crd is being marked for deletion and reconciled cleanly, remove the finalizer.
+		reconcileErr = r.finalize(ctx, crd)
+		if reconcileErr == nil {
+			if crd, err = r.removeFinalizer(ctx, crd); err != nil {
+				logging.FromContext(ctx).Warn("Failed to clear finalizers", zap.Error(err))
+			}
+		}
+	}
+	if reconcileErr != nil {
+		r.Recorder.Eventf(crd, corev1.EventTypeWarning, sourceCRDReconcileFailed, "Source CRD reconciliation failed: %v", reconcileErr)
+	}
+	// Requeue if the reconcile failed.
+	return reconcileErr
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, crd *v1beta1.CustomResourceDefinition) error {
+	// The reconciliation process is as follows:
+	// 	1. Resolve GVR and GVK from a particular Source CRD (i.e., those labeled with duck.knative.dev/source = "true")
+	//  2. Dynamically create a controller for it, if not present already. Such controller is in charge of reconciling
+	//     duckv1.Source resources with that particular GVR..
+
+	gvr, gvk, err := r.resolveGroupVersions(ctx, crd)
+	if err != nil {
+		logging.FromContext(ctx).Error("Error while resolving GVR and GVK", zap.String("CRD", crd.Name), zap.Error(err))
+		return err
+	}
+
+	err = r.reconcileController(ctx, crd, gvr, gvk)
```

Comment:
I think this should be written more like autotrigger but you also want to set up watches on the resolved gvrs. Otherwise you will miss new COs that are created.